### DISCUSSION TITLE
Possible for a public RPC proc to insert to a table that's not publicly writable?

### DISCUSSION THREAD

**Discussion by igorclark** [PostgREST/postgrest#2595] [Q&A] [answered]
Hi there, I'm relatively new to PostgREST so I hope this isn't a dumb question - I've looked around other issues and the docs but haven't found what I'm looking for. If it's there already, please do just point me there - and if this is the wrong place to ask these kinds of questions, please do let me know!

I have a read-only API which has one compute-heavy stored proc. I have a table which I use to store the results of the procedure the first time it gets hit (via `/rpc`) with a particular set of parameters, and on subsequent RPC requests the result is served straight out of the table if the content's already there, so I don't have to do the work again. I don't want that table to be publicly writable, but I do want the RPC to be able to write to it. I want the app to be otherwise read-only for web users, so I only have a `web_anon` user in postgres. So I'm currently making this work by granting `INSERT` to `web_anon` on that cache table, and I've made the join keys in the table foreign keys so anyone trying to mess with it would have to guess valid FKs (as well as the existence of the table itself) - but I just don't like having this open to the world.

Is there a way to allow the RPC to write to the table but not allow any other web visitors to write to it? A config trick, or a different user strategy perhaps?

Thanks!

**Comment by wolfgangwalther** [comment, marked as answer]
Have a look at `SECURITY DEFINER` functions: https://www.postgresql.org/docs/current/sql-createfunction.html.

**Reply by igorclark** [reply]
Super helpful, thank you very much @wolfgangwalther - and for anyone reading, the [Writing SECURITY DEFINER Functions Safely](https://www.postgresql.org/docs/current/sql-createfunction.html#SQL-CREATEFUNCTION-SECURITY) section is particularly germane here. There's also an [article](https://www.cybertec-postgresql.com/en/abusing-security-definer-functions/) I found which is relevant to securing the schema space in the context of security definer functions. Cheers! 👍
